sql/src/main/java/io/crate/planner/node/ddl/UpdateSettingsPlan.java

```diff
@@ -80,32 +82,48 @@ public void executeOrFail(DependencyCarrier dependencies,
+                                                                              params,
+                                                                              subQueryResults);
+
         ClusterUpdateSettingsRequest request = new ClusterUpdateSettingsRequest()
-            .persistentSettings(buildSettingsFrom(persistentSettings, params))
-            .transientSettings(buildSettingsFrom(transientSettings, params));
+            .persistentSettings(buildSettingsFrom(persistentSettings, eval))
+            .transientSettings(buildSettingsFrom(transientSettings, eval));
+
         OneRowActionListener<ClusterUpdateSettingsResponse> actionListener = new OneRowActionListener<>(
             consumer,
             r -> r.isAcknowledged() ? new Row1(1L) : new Row1(0L));
         dependencies.transportActionProvider().transportClusterUpdateSettingsAction().execute(request, actionListener);
     }
 
-    static Settings buildSettingsFrom(Map<String, List<Expression>> settingsMap, Row parameters) {
-        Settings.Builder settings = Settings.builder();
-        for (Map.Entry<String, List<Expression>> entry : settingsMap.entrySet()) {
-            String settingsName = entry.getKey();
+    @VisibleForTesting
+    static Settings buildSettingsFrom(Map<Symbol, List<Symbol>> settingsMap,
+                                      Function<? super Symbol, Object> eval) {
+        Settings.Builder settingsBuilder = Settings.builder();
+        for (Map.Entry<Symbol, List<Symbol>> entry : settingsMap.entrySet()) {
+            String settingsName = eval.apply(entry.getKey()).toString();
+
             if (CrateSettings.isValidSetting(settingsName) == false) {
                 throw new IllegalArgumentException("Setting '" + settingsName + "' is not supported");
             }
+            List<Symbol> value = entry.getValue();
+            CrateSettings.flattenSettings(settingsBuilder, settingsName, eval.apply(Iterables.getOnlyElement(value)));
+        }
 
-            List<Expression> value = entry.getValue();
-            // value can be null to reset settings
-            if (value == null) {
-                settings.put(settingsName, (String) null);
-            } else {
-                CrateSettings.flattenSettings(settings, settingsName,
-                    ExpressionToObjectVisitor.convert(Iterables.getOnlyElement(value), parameters));
-            }
+        Settings settings = settingsBuilder.build();
+        for (String checkForRuntime : settings.keySet()) {
+            CrateSettings.checkIfRuntimeSetting(checkForRuntime);
+        }
+        return settings;
+    }
+
+    private static Map<Symbol, List<Symbol>> buildMap(List<Assignment<Symbol>> assignments) {
+        Map<Symbol, List<Symbol>> result = new HashMap<>();
+        for (Assignment<Symbol> assignment : assignments) {
+            result.put(assignment.columnName(), assignment.expressions());
         }
-        return settings.build();
+        return result;
```

Comment:
What's the reason for this transformation? Wouldn't it be possible to build the settings directly from the assignments?

Reply:
The reason for this transformation is the merge of persistent and transient, which needs to based on the key, not the whole assignment. https://github.com/crate/crate/blob/2fbc9e6e12e2fc27c3a9ecb5dc4da37787cc20cf/sql/src/main/java/io/crate/planner/node/ddl/UpdateSettingsPlan.java#L52